share/doc/wake/basic_wake_tutorial.md

```diff
@@ -0,0 +1,809 @@
+* Using if statement as well as using match function
+```
+def facto1 n = if n == 0 then 1 else  n * facto (n-1)
+```
+
+* facto 5 = `5*4*3*2*1` = 120
+```
+def facto = match _
+  0 = 1
+  n = n * facto (n-1) 
+```
+#### Recursion on Integers
+* Try out these recursive functions and record the result
+```
+def t1 = match _
+  0 = 0
+  n = 2 + t(n-1)
+
+def d1 = match _
+  0 = "de"
+  n = cat ("do",d1(n-1),"da",Nil)
+
+def h n = match n
+  0 = 1
+  n = h(n-1) + h(n-1) 
+
+def j = match _
+  0 = Nil
+  n = (n % 2),j(n / 2)
+
+def m a b = match a b
+  _ 0 = 0
+  a b = a + (m a (b - 1))
+
+def f = match _
+  0 = 0
+  n = 1 - f (n-1)
+
+def g = match _
+  0 = 0
+  n = g (n-1) + 2 * (n-1)
+
+def l = match _
+  0 = 0
+  n = n % 10 + l (n / 10) 
+```
+
+* sumto 4 = 4+3+2+1+0
+```
+def sumto n = match n
+  0 = 0
+  n = n + sumto (n-1)
+```
+			
+* listfrom 4 = (4,3,2,1,Nil)
+```
+def listfrom n = match n
+  0 = Nil
+  n = n , listfrom (n - 1)
+```
+
+* strcopy "ab" 4 = "abababab"
+```
+def strcopy str n = match n 
+  0 = ""
+  n = "{str}{strcopy str (n-1)}"
+```
+
+* listcopy 7 4 = (7,7,7,7,Nil)
+```
+def listcopy l n = match n
+  0 = Nil
+  n = l ,listcopy l (n-1)
+```
+
+* listto 4 = (1,2,3,4,Nil)
+```
+def listto n = listfrom n | reverse
+```
+
+* listodd 7 = (7,5,3,1,Nil)
+```
+def listodd n = if(n<0) then Nil else if(n%2 ==1) then n,listodd (n-2) else n
+```
+
+* sumenes 8 = 8+6+4+2+0 = 20
+```
+def sumevens n = if (n==0) then 0 else if (n%2 ==0) then n + sumevens (n-2) else n
+```
+
+```
+def natstring = "succ"
+def nat n = match n 
+  0 = "zero"
+  n = "{natstring}({nat(n-1)})"
+```
+#### Recursion On Lists
+* Adds all the elements in a list and provids the sum
+```
+def sum n = match n 
+  Nil = 0
+  h,t = h + sum t
+```
+
+* Doubles every element in the list
+```
+def doublelist n = match n
+    Nil = Nil
+    h,t = 2*h, doublelist t
+```
+* The other way of multiplying every element in a list
+```
+def doublelist2 = map (2*_)
+```
+
+* Duplicates every element in a list
+```
+def dupelist n = match n
+    Nil = Nil
+    h,t = (h,h,Nil)++ dupelist t
+```
+
+* Below are the examples which does similar functionality as above
+```
+def dupelist2 = foldr (\x \y x,x,y) Nil
+
+def dupelist1 n = mapFlat (\x (x,x,Nil) ) n
+
+def dupelist3 = mapFlat (\x (x,x,Nil) ) 
+```
+#### More Examples
+* Returns the length of a list
+```
+def length n = len n
+```
+
+* Multiples every element in a list and gives the result
+```
+def prodlist n = foldr (_ * _)1 n
+```
+
+* vallist ("1","2","3","4",Nil) = (1,2,3,4,Nil)
+```
+def vallist n =n | map (\x x | int | getOrElse 0)
+```
+
+* spacelist ("1","2","3",Nil) = ("1","","2","","3","",Nil) 
+```
+def spacelist n = n | mapFlat (\x(x,"",Nil))
+```
+
+* flatten converts list if lists into a list
```

Comment:
is flatten a builtin on lists? can you just put it up in the builtin section?

Reply:
Done

Reply:
Done